backend/de.metas.business/src/main/java/de/metas/invoice/invoiceProcessingServiceCompany/InvoiceProcessingServiceCompanyConfigRepository.java

```diff
@@ -47,77 +45,53 @@
-	configsById = CCache.<InvoiceProcessingServiceCompanyConfigId, InvoiceProcessingServiceCompanyConfig> builder()
-			.tableName(I_InvoiceProcessingServiceCompany.Table_Name)
-			.build();
-
-	private final CCache<Integer, CustomerToConfigAssignmentMap> //
-	customerToConfigAssignmentsCache = CCache.<Integer, CustomerToConfigAssignmentMap> builder()
-			.tableName(I_InvoiceProcessingServiceCompany_BPartnerAssignment.Table_Name)
-			.build();
+	private final CCache<Integer, InvoiceProcessingServiceCompanyConfigMap> configsMapCache =
+			CCache.<Integer, InvoiceProcessingServiceCompanyConfigMap>builder()
+					.tableName(I_InvoiceProcessingServiceCompany.Table_Name)
+					.additionalTableNameToResetFor(I_InvoiceProcessingServiceCompany_BPartnerAssignment.Table_Name)
+					.build();
 
-	public Optional<InvoiceProcessingServiceCompanyConfig> getByCustomerId(@NonNull final BPartnerId customerId)
+	public Optional<InvoiceProcessingServiceCompanyConfig> getByCustomerId(@NonNull final BPartnerId customerId, @NonNull final ZonedDateTime evaluationDate)
 	{
-		return getConfigIdByCustomerId(customerId)
-				.map(this::getById)
-				.filter(InvoiceProcessingServiceCompanyConfig::isActive);
+		final InvoiceProcessingServiceCompanyConfigMap config = configsMapCache.getOrLoad(0, this::retrieveAllCompanyConfigs);
+		return config.getByCustomerIdAndDate(customerId, evaluationDate);
 	}
 
-	private Optional<InvoiceProcessingServiceCompanyConfigId> getConfigIdByCustomerId(@NonNull final BPartnerId customerId)
+	@NonNull
+	private InvoiceProcessingServiceCompanyConfigMap retrieveAllCompanyConfigs()
 	{
-		final CustomerToConfigAssignmentMap customerToConfigAssignmentMap = customerToConfigAssignmentsCache.getOrLoad(0, this::retrieveCustomerToConfigAssignmentMap);
-		return customerToConfigAssignmentMap.getConfigIdByCustomerId(customerId);
-	}
-
-	private InvoiceProcessingServiceCompanyConfig getById(@NonNull final InvoiceProcessingServiceCompanyConfigId configId)
-	{
-		return configsById.getOrLoad(configId, this::retrieveById);
-	}
-
-	private InvoiceProcessingServiceCompanyConfig retrieveById(@NonNull final InvoiceProcessingServiceCompanyConfigId configId)
-	{
-		final I_InvoiceProcessingServiceCompany record = loadOutOfTrx(configId, I_InvoiceProcessingServiceCompany.class);
-
-		return InvoiceProcessingServiceCompanyConfig.builder()
-				.active(record.isActive())
-				.serviceCompanyBPartnerId(BPartnerId.ofRepoId(record.getServiceCompany_BPartner_ID()))
-				.serviceInvoiceDocTypeId(DocTypeId.ofRepoId(record.getServiceInvoice_DocType_ID()))
-				.serviceFeeProductId(ProductId.ofRepoId(record.getServiceFee_Product_ID()))
-				.feePercentageOfGrandTotal(Percent.of(record.getFeePercentageOfGrandTotal()))
-				.build();
+		final ImmutableList<InvoiceProcessingServiceCompanyConfig> collect = queryBL.createQueryBuilder(I_InvoiceProcessingServiceCompany.class)
+				.addOnlyActiveRecordsFilter()
+				.create()
+				.iterateAndStream()
+				.map(record -> {
```

Comment:
extract the mapper into a method (static preferably)

Reply:
Agree with extraction to method, however I don't think it makes sense to have this static, since i'll have to make static `retrieveAllBPartnerDetails`, and that uses `queryBL`. Of course i could just get `queryBL` there, but i don't see a point getting it twice.

Reply:
auch.... that's SQL N+1 problem. u shall avoid it by fetching everything in one round....

Reply:
fixed